IndicoToolkit/Extractions.cs

```diff
@@ -0,0 +1,166 @@
+    public class Extractions {
+        internal List<Prediction> preds = new List<Prediction>();
+        public List<Prediction> removed_predictions = new List<Prediction>();
+        
+        public Extractions(List<Prediction> predictions)
+        {
+            this.preds = predictions;
+        }
+
+        public List<Prediction> Preds() 
+        {
+            return preds;
+        }
+
+        /// <summary>
+        /// Method <c>toDictByLabel</c> generates a dictionary where key is label string and value is list of all predictions of that label.
+        /// </summary>
+        public Dictionary<string, List<Prediction>> toDictByLabel()
+        {
+            Dictionary<string, List<Prediction>> prediction_label_map = new Dictionary<string, List<Prediction>>();
+            for (int i = 0; i < preds.Count; i++) 
+            {
+                Prediction pred = preds[i];
+                string pred_label = pred.getValue("label");
+
+                if (prediction_label_map.ContainsKey(pred_label)) {
+                    prediction_label_map[pred_label].Append(pred);
+                } else {
+                    prediction_label_map.Add(pred_label, new List<Prediction>() {pred});
+                }
+            }
+            return prediction_label_map;
+        }
+        
+        /// <summary>
+        /// Method <c>numPredictions</c> returns number of predictions.
+        /// </summary>
+        public int numPredictions() 
+        {
+            return preds.Count; 
+        }
+
+        /// <summary>
+        /// Method <c>removeByConfidence</c> removes predictions that are less than given confidence.
+        /// </summary>
+        public void removeByConfidence(float confidence = 0.95f, List<string> labels = null) 
+        {
+            List<Prediction> high_conf_preds = new List<Prediction>();
+            for (int i = 0; i < preds.Count; i++) {
+                Prediction pred = preds[i];
+                string pred_label = pred.getValue("label");
+                if (labels != null && !labels.Contains(pred_label)) {
+                    high_conf_preds.Append(pred);
+                } else if (pred.getValue("confidence")[pred_label] >= confidence) {
+                    high_conf_preds.Append(pred);
+                } else {
+                    removed_predictions.Append(pred);
+                }
+            }
+            preds = high_conf_preds;
+        }
+
+        /// <summary>
+        /// Method <c>removeExceptMaxConfidence</c> removes all predictions except the highest confidence within each specified class.
+        /// </summary>
+        public void removeExceptMaxConfidence(List<string> labels) 
+        {
+            /* TODO */
+            return ;
+        }
+        
+        /// <summary>
+        /// Method <c>removeKeys</c> removes specified keys from prediction dictionaries.
+        /// </summary>
+        public void removeKeys(List<string> keys_to_remove = default(List<string>))
```

Comment:
Let's exclude this method for now- would definitely want to make it explicit in the signature as well what the default values are (as opposed to setting them in the if statement)

Reply:
Sounds good. About the default value, c# was complaining, and I found some stacks that say it's not possible to put a non-empty List as a default parameter:

https://stackoverflow.com/questions/31717160/set-default-value-for-parameter-of-liststring-type-in-function
https://stackoverflow.com/questions/6947470/c-how-to-use-empty-liststring-as-optional-parameter#:~:text=It%20is%20impossible.,should%20use%20method%20overloading%20instead.